ts/features/bonus/cgn/screens/merchants/CgnMerchantSearchScreen.tsx

```diff
@@ -0,0 +1,315 @@
+          onCancel={handleCancel}
+          onChangeText={setSearchText}
+          placeholder={I18n.t("bonus.cgn.merchantSearch.input.placeholder")}
+          ref={searchInputRef}
+          value={searchText}
+        />
+      </View>
+      <FlatList
+        ItemSeparatorComponent={() => <Divider />}
+        data={merchants?.length !== 0 ? merchants : undefined}
+        keyExtractor={item => item.id}
+        renderItem={renderItemCallback}
+        ListEmptyComponent={renderListEmptyComponent}
+        keyboardDismissMode={Platform.select({
+          ios: "interactive",
+          default: "on-drag"
+        })}
+        keyboardShouldPersistTaps="handled"
+      />
+    </>
+  );
+}
+
+function MerchantSearchResultListItem({
+  item,
+  searchText
+}: {
+  item: SearchItem;
+  searchText: string;
+}) {
+  const { navigate } =
+    useNavigation<
+      IOStackNavigationProp<CgnDetailsParamsList, "CGN_MERCHANTS_SEARCH">
+    >();
+  return (
+    <View style={IOStyles.horizontalContentPadding}>
+      <ListItemNav
+        onPress={() => {
+          navigate(CGN_ROUTES.DETAILS.MERCHANTS.DETAIL, {
+            merchantID: item.id
+          });
+          Keyboard.dismiss();
+        }}
+        value={
+          <View style={IOStyles.rowSpaceBetween}>
+            <View style={{ flexGrow: 1, flexShrink: 1 }}>
+              <H6>{highlightText({ text: item.name, searchText })}</H6>
+              <Body numberOfLines={2} ellipsizeMode="tail">
+                {highlightText({
+                  text: item.description,
+                  searchText,
+                  esimatedTextLengthToDisplay: item.newDiscounts ? 60 : 100
+                })}
+              </Body>
+            </View>
+            {item.newDiscounts && (
+              <View style={[IOStyles.rowSpaceBetween, IOStyles.alignCenter]}>
+                <Badge
+                  variant="purple"
+                  text={I18n.t("bonus.cgn.merchantsList.news")}
+                />
+              </View>
+            )}
+          </View>
+        }
+        accessibilityLabel={item.name}
+      />
+    </View>
+  );
+}
+
+function highlightText({
+  searchText,
+  text,
+  esimatedTextLengthToDisplay
+}: {
+  text: string;
+  searchText: string;
+  esimatedTextLengthToDisplay?: number;
+}) {
+  const chunks = highlightSearchText({
+    text,
+    searchText,
+    esimatedTextLengthToDisplay
+  });
+  return chunks.map((chunk, index) => (
+    <Text
+      key={index}
+      style={{
+        backgroundColor: chunk.highlighted
+          ? IOColors["turquoise-150"]
+          : undefined
+      }}
+    >
+      {chunk.text}
+    </Text>
+  ));
+}
+
+type HighlightChunk = { highlighted: boolean; text: string };
+
+/**
+ * Highlights search results client side that were made with ILIKE sql operator server side.
+ * Tries to center the first match in the available space if esimatedTextLengthToDisplay provided
+ */
+function highlightSearchText({
```

Comment:
since this function will be used approximately 20 time every time user updates search term, I think this version would create unnecessarily too many object allocations

Reply:
let me know how would you like to proceed

Reply:
> since this function will be used approximately 20 time every time user updates search term, I think this version would create unnecessarily too many object allocations

Yes, makes sense to me